tools/rpm2img

```diff
@@ -56,58 +61,44 @@ VERITY_HASH_ALGORITHM=sha256
-# not update GRUB during our system updates; we would have no way to revert to
-# an earlier copy of the bootloader if it failed to boot.
-#
-# We may eventually want to have an active/passive scheme for EFI partitions,
-# to allow for potential GRUB and shim updates on EFI platforms in cases where
-# we need to deliver security fixes. For now, add a placeholder partition type
-# and reserve space for an alternate bank.
-EFI_BACKUP_TYPECODE="B39CE39C-0A00-B4AB-2D11-F18F8237A21C"
-
-truncate -s 2G "${DISK_IMAGE}"
-# efi: 5M + boot: 40M + root: 920M + hash: 10M + reserved: 25M = 1000M
-# boot partition attributes (-A): 48 = gptprio priority bit; 56 = gptprio successful bit
-# partitions are backwards so that we don't make things inconsistent when specifying a wrong end sector :)
-sgdisk --clear \
-   -n 0:2005M:2047M -c 0:"BOTTLEROCKET-PRIVATE"    -t 0:"${BOTTLEROCKET_PRIVATE_TYPECODE}" \
-   -n 0:1980M:0     -c 0:"BOTTLEROCKET-RESERVED-B" -t 0:"${BOTTLEROCKET_RESERVED_TYPECODE}" \
-   -n 0:1970M:0     -c 0:"BOTTLEROCKET-HASH-B"     -t 0:"${BOTTLEROCKET_HASH_TYPECODE}" \
-   -n 0:1050M:0     -c 0:"BOTTLEROCKET-ROOT-B"     -t 0:"${BOTTLEROCKET_ROOT_TYPECODE}" \
-   -n 0:1010M:0     -c 0:"BOTTLEROCKET-BOOT-B"     -t 0:"${BOTTLEROCKET_BOOT_TYPECODE}" -A 0:"clear":48 -A 0:"clear":56 \
-   -n 0:1005M:0     -c 0:"EFI-BACKUP"              -t 0:"${EFI_BACKUP_TYPECODE}" \
-   -n 0:980M:0      -c 0:"BOTTLEROCKET-RESERVED-A" -t 0:"${BOTTLEROCKET_RESERVED_TYPECODE}" \
-   -n 0:970M:0      -c 0:"BOTTLEROCKET-HASH-A"     -t 0:"${BOTTLEROCKET_HASH_TYPECODE}" \
-   -n 0:50M:0       -c 0:"BOTTLEROCKET-ROOT-A"     -t 0:"${BOTTLEROCKET_ROOT_TYPECODE}" \
-   -n 0:10M:0       -c 0:"BOTTLEROCKET-BOOT-A"     -t 0:"${BOTTLEROCKET_BOOT_TYPECODE}" -A 0:"set":48 -A 0:"set":56 \
-   -n 0:5M:0        -c 0:"EFI-SYSTEM"              -t 0:"${EFI_SYSTEM_TYPECODE}" \
-   -n 0:1M:0        -c 0:"BIOS-BOOT"               -t 0:"${BIOS_BOOT_TYPECODE}" \
-   --sort --print "${DISK_IMAGE}"
+truncate -s "${OS_IMAGE_SIZE_GIB}"G "${OS_IMAGE}"
+
+declare -A partlabel parttype partsize partoff
+set_partition_sizes \
+  "${OS_IMAGE_SIZE_GIB}" "${DATA_IMAGE_SIZE_GIB}" \
+  partsize partoff
+set_partition_labels partlabel
+set_partition_types parttype
+
+declare -a partargs
+for part in \
+  BIOS \
+  EFI-A BOOT-A ROOT-A HASH-A RESERVED-A \
+  EFI-B BOOT-B ROOT-B HASH-B RESERVED-B \
+  PRIVATE ;
+do
+  # Each partition is aligned to a 1 MiB boundary, and extends to the sector
+  # before the next partition starts. Specify the end point in sectors so we
+  # can subtract a sector to fix the off-by-one error that comes from adding
+  # start and size together. (1 MiB contains 2048 512-byte sectors.)
+  part_start=${partoff["${part}"]}
+  part_end=$((${part_start} + ${partsize["${part}"]}))
+  part_end=$((${part_end} * 2048 - 1))
+
+  partargs+=(-n 0:${part_start}M:${part_end})
+  partargs+=(-c 0:${partlabel["${part}"]})
+  partargs+=(-t 0:${parttype["${part}"]})
```

Comment:
```suggestion
  partargs+=(-n 0:"${part_start}M:${part_end}")
  partargs+=(-c 0:"${partlabel["${part}"]}")
  partargs+=(-t 0:"${parttype["${part}"]}")
```
nit.

Reply:
(if you do quote, there's no reason to leave the other constant text out of the quotes, it's more confusing for new people to understand why you're introducing quotes for only some text)

Reply:
I'd prefer to skip the quotes here for consistency with the other uses of the dictionaries. I avoided quoting those expressions because the string keys also need to be in double-quotes, which introduces a fair amount of escaping overhead for relatively little gain.

Reply:
@jpculp explained that this fixes a `shellcheck` warning, which makes sense to me.